substrate-node/pallets/pallet-tfgrid/src/farm.rs

```diff
@@ -1,10 +1,373 @@
+
+        Ok(().into())
+    }
+
+    pub fn _set_farm_dedicated(farm_id: u32, dedicated: bool) -> DispatchResultWithPostInfo {
+        let mut farm = Farms::<T>::get(farm_id).ok_or(Error::<T>::FarmNotExists)?;
+        farm.dedicated_farm = dedicated;
+        Farms::<T>::insert(farm_id, &farm);
+
+        Self::deposit_event(Event::FarmUpdated(farm));
+
+        Ok(().into())
+    }
+
+    pub fn _force_reset_farm_ip(farm_id: u32, ip: Ip4Input) -> DispatchResultWithPostInfo {
+        ensure!(Farms::<T>::contains_key(farm_id), Error::<T>::FarmNotExists);
+        let mut stored_farm = Farms::<T>::get(farm_id).ok_or(Error::<T>::FarmNotExists)?;
+
+        match stored_farm
+            .public_ips
+            .iter_mut()
+            .find(|pubip| pubip.ip == ip)
+        {
+            Some(ip) => {
+                ip.contract_id = 0;
+            }
+            None => return Err(Error::<T>::IpNotExists.into()),
+        };
+
+        Farms::<T>::insert(stored_farm.id, &stored_farm);
+
+        Self::deposit_event(Event::FarmUpdated(stored_farm));
+
+        Ok(().into())
+    }
+
+    pub fn _attach_policy_to_farm(
+        farm_id: u32,
+        limits: Option<FarmingPolicyLimit>,
+    ) -> DispatchResultWithPostInfo {
+        if let Some(policy_limits) = limits {
+            let farming_policy = FarmingPoliciesMap::<T>::get(policy_limits.farming_policy_id);
+            let now = frame_system::Pallet::<T>::block_number();
+
+            // Policy end is expressed in number of blocks
+            if farming_policy.policy_end != T::BlockNumber::from(0 as u32)
+                && now >= farming_policy.policy_created + farming_policy.policy_end
+            {
+                return Err(DispatchErrorWithPostInfo::from(
+                    Error::<T>::FarmingPolicyExpired,
+                ));
+            }
+
+            let mut farm = Farms::<T>::get(farm_id).ok_or(Error::<T>::FarmNotExists)?;
+            // Save the policy limits and farm certification on the Farm object
+            farm.farming_policy_limits = Some(policy_limits.clone());
+            farm.certification = farming_policy.farm_certification;
+            Farms::<T>::insert(farm_id, &farm);
+            Self::deposit_event(Event::FarmUpdated(farm));
+
+            // Give all the nodes in this farm the policy that is attached
+            for node_id in NodesByFarmID::<T>::get(farm_id) {
+                match Nodes::<T>::get(node_id) {
+                    Some(mut node) => {
+                        let current_node_policy =
+                            FarmingPoliciesMap::<T>::get(node.farming_policy_id);
+                        // If the current policy attached to the node is default one, assign it the newly created policy
+                        // because we wouldn't wanna override any existing non-default policies
+                        if current_node_policy.default {
+                            let policy = Self::get_farming_policy(&node)?;
```

Comment:
Because when you attach policy to a farm it comes with some limits 
So maybe this policy will not be assigned to all of its nodes
There is also the concept of default policy in the middle
See `get_farming_policy()` for more details
I confess this one is not easy to understand